### PR TITLE
Fix an out-of-bound scenario in the soft-limiter

### LINKED ISSUE #1023: MT32: std::vector end exceed in void MidiHandler_mt32::Render()->Process...
Source:
git master, lastest commit is
Commit date:	9 hours ago (21.05.2021 00:13:03)
Commit hash:	77b480b1d871456a7338b4d4f0b46c912bfd2b3c

config.h:

```
#define C_DEBUG 1
#define C_HEAVY_DEBUG 1
#define C_FLUIDSYNTH 0

```
System:
Win7 x64, VS2019 (lastest as of 21.5.2021), x64 Debug build run in debugger

dosbox config:

default git config with mt32 config based on mame roms + using model=mt32_old
and core=normal (to ease debugging)

```
mididevice=mt32

[mt32]
romdir=D:\mame_mt32_cm32l_roms\all

# models
# https://github.com/dosbox-staging/dosbox-staging/pull/965#issue-608311889
model=mt32_old

```
Issue:
assert in std::vector "cannot seek vector iterator after end"

Cause:
```
template <int8_t channel>
void SoftLimiter::ScaleOrCopy(const std::vector<float> &in,
                              const uint16_t samples,
                              const float prescalar,
                              const in_iterator_t precross_peak_pos,
                              const in_iterator_t zero_cross_pos,
                              const float global_peak,
                              const float tail,
                              std::vector<int16_t> &out)
{
	assert(samples >= 2); // need at least one frame
	auto in_start = in.begin() + channel;
	const auto in_end = in.begin() + channel + samples; <-- exceeds vector end by 1
```

happens on startup of dosbox

source
stack
assert

### PR DESCRIPTION
**Pull request by kcgen**
VisualStudio 2019 flags the second-channels end-offset as invalid (which is correct), however the offset is used as a `!=` boundary condition, and never actually read or written to.

This PR fixes this by using a `<` condition, and also only setting the end-offset to the last sample (at most).  

Fixes #1023.

### CODE REVIEW

**Comment by LowLevelMahn** [comment]
!= is more idiomatic than <

i think your fix should work without changing !=

**Comment by kcgen** [comment]
> != is more idiomatic than < (or better said: it looks strange with iterators)

I agree :-)

> i think your fix should work without changing !=

Previously, the left and right iterators were both stepped forward by the 'channel' value (0 for left, 1 for right). 

So for a single frame sequence, or simply two samples (one left, one right), we would have:

Left start = 0 // valid
Left end = 2 // knowingly beyond, passes assert (::end)
Right start = 1 // valid
Right end = 3 // knowingly beyond, fails assert (::end + 1)

The processing code is channel-agnostic, and steps the iterator forward by two and breaks when we hit end.

Left start + 2 is the left end
Right start +2 is the right end

This PR now caps both left and right ends to the same legal end offset.

Left start = 0 // valid
Left end = 2 // knowingly beyond, passes assert
Right start = 1 // valid
Right end = 2 // knowingly beyond, passes assert

The problem is when we march the iterators forward by two, the right will no longer be in-sync with the end and walk beyond:

Left start + 2 is the left end
Right start +2 is one beyond right end, so keeps going (out of bounds).

**Comment by LowLevelMahn** [comment]
it feels not right that your algorithm carries around or produces invalid iterators

**Comment by kcgen** [comment]
> it feels not right that your algorithm carries around or produces invalid iterators

After this PR:

This value:

``` c++
const auto in_end = in.begin() + samples;
```

Will, at most, be `vector::end()` for both the left and right audio channels (see: `https://www.cplusplus.com/reference/vector/vector/end/`).

Unless you are referring somewhere else?

**Comment by kcgen** [comment]
Does this PR still produce the issue you previously reported?

**Comment by LowLevelMahn** [comment]
> Does this PR still produce the issue you previously reported?

will check tomorrow

**Comment by LowLevelMahn** [comment]
same assert - different place

```
void SoftLimiter::LinearScale(in_iterator_t in_pos,
                              const in_iterator_t in_end,
                              out_iterator_t out_pos,
                              const float scalar) const noexcept
{
	while (in_pos < in_end) {
		const auto scaled = (*in_pos) * scalar;
		assert(fabsf(scaled) < INT16_MAX);
		*out_pos = static_cast<int16_t>(scaled);
		out_pos += 2;
		in_pos += 2; <--- asserts here
```

image

there are just too many invalid iterators around - i mean - moren then 0 :)

switching over to raw-pointers will not help here - could also hide the error even more from ASAN

**Comment by kcgen** [comment]
Thanks for this suggestion! Will try the functor and lamda solution tomorrow.
Regarding the last question, " this is also a candiate for invalid iterators", it's not - it will always land on ::end evenly, and can continue using `!=` break criteria (I had only changed it to make it follow the overall less than style.. but it wasn't needed).

**Comment by LowLevelMahn** [comment]
> Thanks for this suggestion! Will try the functor and lamda solution tomorrow.

and you should also consider making more of your private methods static if they use no members, and maybe remove them from the header and convert them to cpp-only implementations

> Regarding the last question, " this is also a candiate for invalid iterators", it's not - it will always land on ::end evenly and can continue using the `!=` break criteria (I had only changed it to use the less than criteria to be consistent in style.. but this wasn't needed).

yea, != would look less suspicious

**Comment by kcgen** [comment]
The distance-lamba approach doesn't populate the last sample in the right channel, so produces popping.

``` text
./build/clang-both-a_u/tests/soft_limiter
Running main() from /build/googletest-OI1WhS/googletest-1.10.0.20201025/googletest/src/gtest_main.cc
[==========] Running 18 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 18 tests from SoftLimiter
[ RUN      ] SoftLimiter.InboundsProcessAllFrames
../../tests/soft_limiter.cpp:36: Failure
Expected equality of these values:
  out
    Which is: { -3, -2, -1, 0, 1, 0 }
  expected
    Which is: { -3, -2, -1, 0, 1, 2 }
[  FAILED  ] SoftLimiter.InboundsProcessAllFrames (0 ms)
[ RUN      ] SoftLimiter.InboundsProcessPartialFrames
../../tests/soft_limiter.cpp:49: Failure
Expected equality of these values:
  out[1]
    Which is: 0
  expected[1]
    Which is: -2
[  FAILED  ] SoftLimiter.InboundsProcessPartialFrames (0 ms)
[ RUN      ] SoftLimiter.InboundsProcessTooManyFrames
[       OK ] SoftLimiter.InboundsProcessTooManyFrames (2 ms)
[ RUN      ] SoftLimiter.OutOfBoundsLeftChannel
../../tests/soft_limiter.cpp:71: Failure
Expected equality of these values:
  out
    Which is: { -4, 32000, 32766, 32000, 2, 0 }
  expected
    Which is: { -4, 32000, 32766, 32000, 2, 32000 }
[  FAILED  ] SoftLimiter.OutOfBoundsLeftChannel (0 ms)
[ RUN      ] SoftLimiter.OutOfBoundsRightChannel
../../tests/soft_limiter.cpp:84: Failure
Expected equality of these values:
  out
    Which is: { 32000, -1, 32000, 32765, 32000, 0 }
  expected
    Which is: { 32000, -1, 32000, 32765, 32000, 2 }
[  FAILED  ] SoftLimiter.OutOfBoundsRightChannel (0 ms)
[ RUN      ] SoftLimiter.OutboundsBothChannelsPositive
../../tests/soft_limiter.cpp:97: Failure
Expected equality of these values:
  out
    Which is: { -4, -1, 32766, 32765, 2, 0 }
  expected
    Which is: { -4, -1, 32766, 32765, 2, 2 }
[  FAILED  ] SoftLimiter.OutboundsBothChannelsPositive (0 ms)
[ RUN      ] SoftLimiter.OutboundsBothChannelsNegative
../../tests/soft_limiter.cpp:110: Failure
Expected equality of these values:
  out
    Which is: { -4, -1, -32766, -32765, 2, 0 }
  expected
    Which is: { -4, -1, -32766, -32765, 2, 2 }
[  FAILED  ] SoftLimiter.OutboundsBothChannelsNegative (0 ms)
[ RUN      ] SoftLimiter.OutboundsBothChannelsMixed
../../tests/soft_limiter.cpp:124: Failure
Expected equality of these values:
  out
    Which is: { 19999, -13332, 32766, -32766, 19999, 0 }
  expected
    Which is: { 19999, -13332, 32766, -32766, 19999, -13332 }
[  FAILED  ] SoftLimiter.OutboundsBothChannelsMixed (0 ms)
[ RUN      ] SoftLimiter.OutboundsBigOneReleaseStep
../../tests/soft_limiter.cpp:140: Failure
Expected equality of these values:
  out
    Which is: { -17920, 0 }
  expected
    Which is: { -17920, 13434 }
[  FAILED  ] SoftLimiter.OutboundsBigOneReleaseStep (0 ms)
[ RUN      ] SoftLimiter.OutboundsBig600ReleaseSteps
../../tests/soft_limiter.cpp:156: Failure
Expected equality of these values:
  out
    Which is: { -32766, 0 }
  expected
    Which is: { -32766, 32766 }
[  FAILED  ] SoftLimiter.OutboundsBig600ReleaseSteps (0 ms)
[ RUN      ] SoftLimiter.OutboundsSmallTwoReleaseSteps
../../tests/soft_limiter.cpp:171: Failure
Expected equality of these values:
  out
    Which is: { -32766, 0 }
  expected
    Which is: { -32766, 32766 }
[  FAILED  ] SoftLimiter.OutboundsSmallTwoReleaseSteps (0 ms)
[ RUN      ] SoftLimiter.OutboundsSmallTenReleaseSteps
../../tests/soft_limiter.cpp:187: Failure
Expected equality of these values:
  out
    Which is: { -32766, 0 }
  expected
    Which is: { -32766, 32766 }
[  FAILED  ] SoftLimiter.OutboundsSmallTenReleaseSteps (0 ms)
[ RUN      ] SoftLimiter.OutboundsPolyJoinPositive
../../tests/soft_limiter.cpp:201: Failure
Expected equality of these values:
  out
    Which is: { 18000, 18000, 20000, 20000, 22000, 0 }
  expected_first
    Which is: { 18000, 18000, 20000, 20000, 22000, 22000 }
../../tests/soft_limiter.cpp:209: Failure
Expected equality of these values:
  out
    Which is: { 24266, 16383, 32766, 32766, 16383, 0 }
  expected_second
    Which is: { 24266, 24266, 32766, 32766, 16383, 16383 }
[  FAILED  ] SoftLimiter.OutboundsPolyJoinPositive (1 ms)
[ RUN      ] SoftLimiter.OutboundsPolyJoinNegative
../../tests/soft_limiter.cpp:223: Failure
Expected equality of these values:
  out
    Which is: { -18000, -18000, -20000, -20000, -22000, 0 }
  expected_first
    Which is: { -18000, -18000, -20000, -20000, -22000, -22000 }
../../tests/soft_limiter.cpp:231: Failure
Expected equality of these values:
  out
    Which is: { -24266, -16383, -32766, -32766, -16383, 0 }
  expected_second
    Which is: { -24266, -24266, -32766, -32766, -16383, -16383 }
[  FAILED  ] SoftLimiter.OutboundsPolyJoinNegative (0 ms)
[ RUN      ] SoftLimiter.OutboundsJoinWithZeroCross
../../tests/soft_limiter.cpp:253: Failure
Expected equality of these values:
  out
    Which is: { 12287, 1000, 20478, 1000, 26212, 1000, 0, 1000, -12287, 1000, -32765, 0 }
  expected_second
    Which is: { 12287, 1000, 20478, 1000, 26212, 1000, 0, 1000, -12287, 1000, -32765, 1000 }
../../tests/soft_limiter.cpp:263: Failure
Expected equality of these values:
  out
    Which is: { -20524, 1000, -12314, 1000, -8209, 1000, -4104, 1000, 0, 1000, 2462, 0 }
  expected_third
    Which is: { -20524, 1000, -12314, 1000, -8209, 1000, -4104, 1000, 0, 1000, 2462, 1000 }
[  FAILED  ] SoftLimiter.OutboundsJoinWithZeroCross (0 ms)
[ RUN      ] SoftLimiter.ScaleAttenuate
../../tests/soft_limiter.cpp:274: Failure
Expected equality of these values:
  out
    Which is: { -30000, 0 }
  expected_first
    Which is: { -30000, 30000 }
../../tests/soft_limiter.cpp:283: Failure
Expected equality of these values:
  out
    Which is: { -15000, 0 }
  expected_scaled
    Which is: { -15000, 3000 }
[  FAILED  ] SoftLimiter.ScaleAttenuate (0 ms)
[ RUN      ] SoftLimiter.ScaleAmplify
../../tests/soft_limiter.cpp:294: Failure
Expected equality of these values:
  out
    Which is: { -10000, 0 }
  expected_first
    Which is: { -10000, 10000 }
../../tests/soft_limiter.cpp:303: Failure
Expected equality of these values:
  out
    Which is: { -15000, 0 }
  expected_scaled
    Which is: { -15000, 11000 }
[  FAILED  ] SoftLimiter.ScaleAmplify (0 ms)
[ RUN      ] SoftLimiter.RangeMultiply
../../tests/soft_limiter.cpp:319: Failure
Expected equality of these values:
  out
    Which is: { -20000, 0 }
  expected
    Which is: { -20000, 20000 }
[  FAILED  ] SoftLimiter.RangeMultiply (0 ms)
[----------] 18 tests from SoftLimiter (3 ms total)

[----------] Global test environment tear-down
[==========] 18 tests from 1 test suite ran. (3 ms total)
[  PASSED  ] 1 test.
[  FAILED  ] 17 tests, listed below:
[  FAILED  ] SoftLimiter.InboundsProcessAllFrames
[  FAILED  ] SoftLimiter.InboundsProcessPartialFrames
[  FAILED  ] SoftLimiter.OutOfBoundsLeftChannel
[  FAILED  ] SoftLimiter.OutOfBoundsRightChannel
[  FAILED  ] SoftLimiter.OutboundsBothChannelsPositive
[  FAILED  ] SoftLimiter.OutboundsBothChannelsNegative
[  FAILED  ] SoftLimiter.OutboundsBothChannelsMixed
[  FAILED  ] SoftLimiter.OutboundsBigOneReleaseStep
[  FAILED  ] SoftLimiter.OutboundsBig600ReleaseSteps
[  FAILED  ] SoftLimiter.OutboundsSmallTwoReleaseSteps
[  FAILED  ] SoftLimiter.OutboundsSmallTenReleaseSteps
[  FAILED  ] SoftLimiter.OutboundsPolyJoinPositive
[  FAILED  ] SoftLimiter.OutboundsPolyJoinNegative
[  FAILED  ] SoftLimiter.OutboundsJoinWithZeroCross
[  FAILED  ] SoftLimiter.ScaleAttenuate
[  FAILED  ] SoftLimiter.ScaleAmplify
[  FAILED  ] SoftLimiter.RangeMultiply
```

**Comment by kcgen** [comment]
I suspect it can be made to work by adding 'channel' (which is one for the right-channel) to the step-count, but then the iterators for the last pass will be incremented out of bounds (but not accessed; and we're back to the current situation).

**Comment by LowLevelMahn** [comment]
> The distance-lamba approach doesn't populate the last sample in the right channel, so produces popping.

it just sounded ok for me, did not test it in depth, nor did i know of the test

is the test part of the standard vs2019 build or do i need extra build settings?

**Comment by kcgen** [comment]
> it just sounded ok for me, did not test it in depth

It's subtle - the limiter processes audio in "large" batches, and the zero'ing only happens on the last sample in the right channel - and it's zeroed - so depending on how that sounds, it can simply be silence (1/44100th or 1/48000th of a second of silence - so could be quite inaudible!)

> is the test part of the standard vs2019 build or do i need extra build settings?

I don't think it's part of the Visualstudio build (yet) - however it uses the google test framework.  If you have Linux or macOS (or a VM), the steps are very easy:

https://github.com/dosbox-staging/dosbox-staging/blob/master/README.md#linux-macos

The line that says `meson setup ..` produces an optimized release build, however in our case to produce a debug build (plus sanitizer instrumentation), use this:

``` shell
meson setup -Dbuildtype=debugoptimized -Db_sanitize=address,undefined -Db_lundef=false -Dc_args=-fsanitize-recover=all -Dcpp_args=-fsanitize-recover=all build
```

The test executables will be in your `build/tests` directory and can be directly (or pass `--help` to see options).

Thanks for the help on this!

**Comment by LowLevelMahn** [comment]
> ```shell
> meson setup -Dbuildtype=debugoptimized -Db_sanitize=address,undefined build
> ```

maybe that works also under windows/vs2019, i really want the checking microsoft stl behind
or i just add the testcode directly to the sln

**Comment by kcgen** [comment]
> I really want the checking microsoft STL behind
I agree; the combination of the STL checking plus unit test results would give you a direct assessment of both aspects.

**Comment by kcgen** [comment]
Closing this PR because it doesn't solve the issue.  @LowLevelMahn, we can keep discussing approaches to solve it here though.

I originally used basic integer indexes, however switched to iterators because it made the code more direct and involved a bit fewer additions. In light of this problem though, it probably makes sense to move back to int indexes. 

Curious if you've had any luck with the unit tests in vs2019?

**Comment by LowLevelMahn** [comment]
> Curious if you've had any luck with the unit tests in vs2019?

meson does not run under my Win7 Testmachine :/

i tried to add a console project + the test files + gtest -> compiles but does not link to due missing gtest macro related code

`example.obj : error LNK2019: Verweis auf nicht aufgelöstes externes Symbol ""unsigned char __cdecl drive_index(char)" (?drive_index@@YAED@Z)" in Funktion ""private: virtual void __cdecl `anonymous namespace'::DriveIndexTest_BigLetters_Test::TestBody(void)" (?TestBody@DriveIndexTest_BigLetters_Test@?A0xe98af231@@EEAAXXZ)".`

it would be great if the test-code could be also available in the vs-solution - so everyone is able to run the unit-test out of the box

**Comment by kcgen** [comment]
I attempted to use GCC's checked STL library (libstdc++-dbg), per the discussion here: https://gcc.gnu.org/onlinedocs/libstdc++/manual/debug_mode.html

Installed the debug library:

``` shell
sudo apt install libstdc++6-10-dbg
echo /usr/lib/x86_64-linux-gnu/debug | sudo tee /etc/ld.so.conf.d/debug.conf
sudo ldconfig
```

Configured Staging to use the Debug library:

meson setup -Dbuildtype=debug -Dcpp_args=-D_GLIBCXX_DEBUG -Dunit_tests=disabled  

Confirmed the binary is linked with the debug binary: ` ldd /path/to/dosbox`

Launched the executable... and .. it immediately triggers on launch citing an attempt to _"increment a singular iterator"_.

This is in DOSBox's conf parser, just as it's coming alive, well before any emulation even starts to run.

``` c++
/usr/include/c++/10/debug/safe_iterator.h:328:
In function:
    __gnu_debug::_Safe_iterator<_Iterator, _Sequence, _Category>& 
    __gnu_debug::_Safe_iterator<_Iterator, _Sequence, 
    _Category>::operator++() [with _Iterator = 
    std::__cxx1998::_Deque_iterator<Section::Function_wrapper, 
    Section::Function_wrapper&, Section::Function_wrapper*>; _Sequence = 
    std::__debug::deque<Section::Function_wrapper>; _Category = 
    std::forward_iterator_tag]

Error: attempt to increment a singular iterator.

Objects involved in the operation:
    iterator "this" @ 0x0x7ffd9e7166f0 {
      type = std::__cxx1998::_Deque_iterator<Section::Function_wrapper, Section::Function_wrapper&, Section::Function_wrapper*> (mutable iterator);
      state = singular;
      references sequence with type 'std::__debug::deque<Section::Function_wrapper, std::allocator<Section::Function_wrapper> >' @ 0x0x558bf9687d70
    }
Aborted
```

The same conf parse code passes:
 - warning checks at compile-time,
 - static analyzer checks by Clang Analysis, Coverity Scan, PVS Studio, and LGTM,
 - instrumented runtime address and undefined-behavior checks, and
 - runtime Valgrind memory checks

So my guess is these debug asserts are even more sensitive than those in visual studio - given on your side you mention actually being able to run and hear music.

**Comment by kcgen** [comment]
> i tried to add a console project + the test files + gtest -> compiles but does not link to due missing gtest macro related code, and some WIN32 defines were not properly set

I came across some notes here on how to get GoogleTest running on Visual Studio - although it's 2012 not 2019:
https://www.bogotobogo.com/cplusplus/google_unit_test_gtest.php

(sorry if this is all duplicate though; from the looks of it's you've tried hard to get it working too). 

> it would be great if the test-code could be also available in the vs-solution - so everyone is able to run the unit-test out of the box

No disagreement here :-)  It's just this is not our working dev and test environment; @dreamer and I both use, dev, and test on Linux.  

Google Test also runs in our CI environment with just an extra couple commands. 

PR's are welcome for anyone who wants to get Google Test up and running in the VS Solution; and we could also launch those Windows-tests under Windows CI. 
 
> would also crash the windows build if is there is a windows/VS2019 test-run possible for the unit-test on github/actions?

Sorry, I'm not following your question - can you rephrase it?

**Comment by LowLevelMahn** [comment]
> I attempted to use GCC's checked STL library (libstdc++-dbg), per the discussion here: 
> ...
> Launched the executable... and .. it immediately triggers on launch citing an attempt to _"increment a singular iterator"_.
> ...
> So my guess is these debug asserts are even more sensitive than those in visual studio - given on your side you mention actually being able to run and hear music.

its sad to see how many differences platform/compilers and standard libs one need to throw at code for bug detection :)

>So my guess is these debug asserts are even more sensitive than those in visual studio
>given on your side you mention actually being able to run and hear music.

better said: maybe senstitive in the same or/and other areas

**Comment by LowLevelMahn** [comment]
> I came across some notes here on how to get GoogleTest running on Visual Studio - although they're written assuming VS2012 and not VS2019:
> https://www.bogotobogo.com/cplusplus/google_unit_test_gtest.php

thank for the link - gtest runs for shure under VS - i once was part of a project using getest in VS

do i just add the stuff to the VS-solution or is meson (i never used meson before) responsible (like CMake do) to create the solution file?

> No disagreement here :-) It's just this is not our working dev and test environment; @dreamer and I both use, dev, and test on Linux.

as you can see - the Windows/VS platform also helps to find more bugs :)
(and i think its still the primary platform for dosbobx usage)




> PR's are welcome for anyone who wants to get Google Test up and running in the VS Solution; and we could also launch those Windows-tests under Windows CI.
> 
> > would also crash the windows build if is there is a windows/VS2019 test-run possible for the unit-test on github/actions?
> 
> Sorry, I'm not following your question - can you rephrase it?

statement answered my question: you ARE able to run the unit-test under a Windows CI, and the VS2019 STL assert would happen there

**Comment by LowLevelMahn** [comment]
why is your RWQeue splitted into h and cpp file - when its completely template based - is that even C++ standard? i don't see where your header reaches directly the code in the cpp

**Comment by LowLevelMahn** [comment]
the unit-test build and runst (not without fails) inside VS2019

what i've done/got so far

added a empty project to the solution
-wide char
-added WIN32 to postprocessor settings
-install gtest with vcpkg
-added gmock_maind.lib to input libs
-added include path to dosbox/include
-refered all the cpps that contains needed source (from the test itself and dosbox)

problems:
-still not compatible with HEAVY_DEBUG - needs debug.cpp->memory.cpp->.... disabled HEAVY-DEBUG (for me) for now
-VS don't likes multiple cpps files with same name (unit-test cpps got the same name as test source implementation) - can we prefix the tests source with test_ or something?

but it think 90% ready :)

**Comment by kcgen** [comment]
You're making excellent progress on this!

1. Regarding the same-named unit test files, it's mentioned here: https://stackoverflow.com/a/22936230

    They say to set the Project-wide settings: Configurtion Properties->C/C++->Output Files to `$(IntDir)%(RelativeDir)` or even better ` $(IntDir)\%(RelativeDir)\%(Filename).obj`

    Unfortunately I wasn't able to follow that answer, so could not test the solution on my side. I simply don't know where to find that setting; here's my failed attempted: 
    

    If you can figure it out, can you check if the setting is applied inside epository's `vs/dosbox.sln` or `vs/dosbox.vcxproj` files?  If yes, the we can add this so everyone can benefit. But if it's more of a user-settings (outside of the repo), then I suggest we rename the test cpps names.  I would use `<name>_tests.cpp`

 2. Regarding the progress you've made (creating a new project inside the solution, etc.. ), if you have updated .sln or other VS files, can you make a branch and PR it or even just attach them here? 

 3. Regarding steps "outside" the repository such as installing `gtest` with vcpkg - this is great! please keep all those notes so we can document them in the BUILD.md notes.

4. Regarding getting the gtest libs inside vcpkg automatically, and how to copy gtest.dll into the output folder?

    Does `vcpkg integrate install` fix it?  I'm looking at: https://github.com/Microsoft/vcpkg/blob/master/docs/users/integration.md#user-wide-for-msbuild-recommended-for-open-source-msbuild-projects

    This page: https://www.debugcn.com/en/article/15901742.html shows how to ensure `gtest_main.lib` is linked.    
    They refer to finding these GUI settings and changing Configuration Properties > Linker > Input > Additional Dependencies. 

    I sure hope we can codify these inside the `sln` or `vcxprog` files so everyone can benefit from the get-go.

**Comment by kcgen** [comment]
@LowLevelMahn , I've split this into a new issue here: https://github.com/dosbox-staging/dosbox-staging/issues/1028.

My reply is below the first post.

Let's keep progressing it under there. :slightly_smiling_face:
